src/job_queue/job_result_wrapper.rs

```diff
@@ -0,0 +1,267 @@
+use super::traits::JobResult;
+use super::JobCompletion;
+
+/// A generic wrapper around a job-specific result type `T` that implements the
+/// [`JobResult`] trait.
+///
+/// This wrapper simplifies the process of:
+///
+/// * Returning concrete job results (`T`) as trait objects (`Box<dyn JobResult>`).
+/// * Attempting to convert a `Box<dyn JobResult>` back into the original concrete type `T`.
+///
+/// # Type Parameters
+///
+/// * `T`: The specific type of the job result being wrapped. This type must be
+///   `'static`, `Send`, and `Sync`.
+///
+/// `JobResultWrapper` also implements the following traits if T
+/// implements the trait:
+///   Debug, Clone, Copy, Display, PartialOrd, Ord, PartialEq, Eq,
+pub struct JobResultWrapper<T>(T);
+
+impl<T: 'static + Send + Sync> JobResult for JobResultWrapper<T> {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn into_any(self: Box<Self>) -> Box<dyn Any> {
+        self
+    }
+}
+
+impl<T> Deref for JobResultWrapper<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for JobResultWrapper<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<T: 'static> TryFrom<JobCompletion> for JobResultWrapper<T> {
+    type Error = JobHandleError;
+
+    fn try_from(job_completion: JobCompletion) -> Result<Self, Self::Error> {
+        JobResultWrapper::try_from_completion(job_completion)
+    }
+}
+
+impl<'a, T: 'static> TryFrom<&'a JobCompletion> for &'a JobResultWrapper<T> {
+    type Error = JobHandleError;
+
+    fn try_from(job_completion: &'a JobCompletion) -> Result<Self, Self::Error> {
+        JobResultWrapper::try_from_completion_ref(job_completion)
+    }
+}
+
+impl<T: 'static> TryFrom<Box<dyn JobResult>> for JobResultWrapper<T> {
+    type Error = JobHandleError;
+
+    fn try_from(job_result: Box<dyn JobResult>) -> Result<Self, Self::Error> {
+        JobResultWrapper::try_from_boxed_job_result(job_result)
+    }
+}
+
+impl<'a, T: 'static> TryFrom<&'a dyn JobResult> for &'a JobResultWrapper<T> {
+    type Error = JobHandleError;
+
+    fn try_from(job_result: &'a dyn JobResult) -> Result<Self, Self::Error> {
+        JobResultWrapper::try_from_boxed_job_result_ref(job_result)
+    }
+}
+
+impl<T: 'static + Send + Sync> From<JobResultWrapper<T>> for Box<dyn JobResult> {
+    fn from(wrapper: JobResultWrapper<T>) -> Self {
+        Box::new(wrapper) as Box<dyn JobResult>
+    }
+}
+
+impl<T: Debug> Debug for JobResultWrapper<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("JobResultWrapper").field(&self.0).finish()
+    }
+}
+
+impl<T: Display> Display for JobResultWrapper<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+// Clippy warns about a manual Clone impl on a Copy type, but this is necessary
+// to handle the case where T is Clone but not Copy.
```

Comment:
This is not true: the trait bound on the generic(s) dictate(s) what is (and is not) derived. [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=445affd5d593c24643fbff89be9f6daa)

Also applies to various other manual trait implementations below.

Reply:
Yeah, it's definitely an oddity of rust. For example, another case where this trips people up is:

```rust
#[derive(Clone)]
struct Foo<T>(Arc<T>); // could be Clone no matter the T, but isn't
```

There are crates like [`derive-where`](https://crates.io/crates/derive-where) that help out in such cases.